### PR TITLE
feat: add segment upload RESTful API for indexer gateway

### PR DESCRIPTION
**Pull request by wanliqun**
- Implement `FileSegmentUploader` for parallel batch file segment uploads
- Refactor common segment calculation logic for reuse and clarity

- - -
This change is [Reviewable](https://reviewable.io/reviews/0glabs/0g-storage-client/64)

### CODE REVIEW

File: cmd/indexer.go [hunk 1]
```diff
@@ -44,6 +45,7 @@ func init() {
 	indexerCmd.Flags().IntVar(&indexerArgs.locationCache.CacheSize, "file-location-cache-size", 100000, "size of file location cache")
 
 	indexerCmd.Flags().Uint64Var(&indexerArgs.maxDownloadFileSize, "max-download-file-size", 100*1024*1024, "Maximum file size in bytes to download")
+	indexerCmd.Flags().UintVar(&indexerArgs.ExpectedReplica, "expected-replica", 2, "Expected number of replications to upload")
```
Comment (MiniFrenchBread) [inline cmd/indexer.go:48]: 2 for default value is too strict, 1 is fine

File: indexer/gateway/download.go [hunk 2]
```diff
@@ -141,15 +141,28 @@ func downloadFileInFolder(c *gin.Context) {
...
-func getFileInfo(ctx context.Context, root string, txSeq uint64) (*node.FileInfo, error) {
+func getFileInfo(ctx context.Context, root string, txSeq uint64) (info *node.FileInfo, err error) {
 	if len(clients) == 0 {
 		return nil, errors.New("no clients available")
 	}
 
-	if root != "" {
-		return clients[0].GetFileInfo(ctx, common.HexToHash(root))
+	for _, client := range clients {
+		if root != "" {
```
Comment (MiniFrenchBread) [inline indexer/gateway/download.go:150]: change txSeq type to a pointer and try to get file info by txSeq first

File: indexer/gateway/upload.go [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+	"github.com/0glabs/0g-storage-client/transfer"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/gin-gonic/gin"
+	"github.com/pkg/errors"
+)
+
+var expectedReplica uint
+
+type UploadSegmentRequest struct {
+	Root  string          `form:"root" json:"root" binding:"required"`   // File merkle root
```
Comment (MiniFrenchBread) [inline indexer/gateway/upload.go:21]: Can we replace the root with txSeq, or add a new field for txSeq here?

In sharding scenario, there may be files with same root but different txSeq, their starting positions on the flow may also be different, which makes the segments that the same node needs to store for these identical files different for each one.

**Review by MiniFrenchBread — CHANGES_REQUESTED** [review]

**Review by wanliqun — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/0glabs/0g-storage-client/64)* status: 0 of 10 files reviewed, 3 unresolved discussions (waiting on @boqiu and @MiniFrenchBread)

___
*[`cmd/indexer.go` line 48 at r1](https://reviewable.io/reviews/0glabs/0g-storage-client/64#gh-1802884946:-O9OMCAQDatylP8v1XQh:b-896fix) ([raw file](https://github.com/0glabs/0g-storage-client/blob/eb68b4127d7f2fd689b3607c914a7d8373aa0d8c/cmd/indexer.go#L48)):*
<details><summary><i>Previously, MiniFrenchBread wrote…</i></summary><blockquote>

2 for default value is too strict, 1 is fine
</blockquote></details>

Done.
___
*[`indexer/gateway/download.go` line 150 at r1](https://reviewable.io/reviews/0glabs/0g-storage-client/64#gh-1802892691:-O9OMFApEMRswgNHzlJR:b-896fix) ([raw file](https://github.com/0glabs/0g-storage-client/blob/eb68b4127d7f2fd689b3607c914a7d8373aa0d8c/indexer/gateway/download.go#L150)):*
<details><summary><i>Previously, MiniFrenchBread wrote…</i></summary><blockquote>

change txSeq type to a pointer and try to get file info by txSeq first
</blockquote></details>

Done.
___
*[`indexer/gateway/upload.go` line 21 at r1](https://reviewable.io/reviews/0glabs/0g-storage-client/64#gh-1802902082:-O9OMHwY0DV5w9BZH28-:b-896fix) ([raw file](https://github.com/0glabs/0g-storage-client/blob/eb68b4127d7f2fd689b3607c914a7d8373aa0d8c/indexer/gateway/upload.go#L21)):*
<details><summary><i>Previously, MiniFrenchBread wrote…</i></summary><blockquote>

Can we replace the root with txSeq, or add a new field for txSeq here?

In sharding scenario, there may be files with same root but different txSeq, their starting positions on the flow may also be different, which makes the segments that the same node needs to store for these identical files different for each one.
</blockquote></details>

Done.

**Review by MiniFrenchBread — APPROVED** [review]

**Pull request merged**